generator/internal/parser/httprule/http_rule_parser.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httprule
+
+import (
+	"fmt"
+	"github.com/googleapis/google-cloud-rust/generator/internal/api"
+)
+
+// The logic in this file is based on the Mustache template implementation.
+// Reference:
+// - https://go.dev/talks/2011/lex.slide (slides)
+// - https://go.dev/talks/2011/lex/r59-lex.go (code)
+
+func Parse(pathTemplate string) ([]api.PathSegment, error) {
+
+	l := &lexer{
+		input:    pathTemplate,
+		state:    slashState, // the first state is always slashState
+		segments: make(chan segment, 2),
+	}
+	var segments []api.PathSegment
+	for {
+		s := l.nextItem()
+		switch s.typ {
+		case segmentLiteral:
+			segments = append(segments, api.NewLiteralPathSegment(s.val))
+		case segmentIdentifier:
+			segments = append(segments, api.NewFieldPathPathSegment(s.val))
+		case segmentVerb:
+			segments = append(segments, api.NewVerbPathSegment(s.val))
+		case segmentError:
+			return nil, fmt.Errorf("error parsing path template (%s): %s", pathTemplate, s.val)
+		case segmentEOF:
+			return segments, nil
+		}
+
+	}
+}
+
+// ### Path template syntax
+//
+//	Template = "/" Segments [ Verb ] ;
+//	Segments = Segment { "/" Segment } ;
+//	Segment  = "*" | "**" | LITERAL | Variable ;
+//	Variable = "{" FieldPath [ "=" Segments ] "}" ;
+//	FieldPath = IDENT { "." IDENT } ;
+//	Verb     = ":" LITERAL ;
+const (
+	segmentError segmentType = iota // 0
+	segmentLiteral
+	segmentIdentifier
+	segmentVerb //a verb is actually ':' + LITERAL,
+	// but we need a way to differentiate a literal within a segment, from a literal within a verb
+	segmentEOF
+)
+
+const (
+	slash    = '/'
+	star     = '*'
+	varLeft  = '{'
+	varRight = '}'
+	varSep   = '='
+	verbSep  = ':'
+)
+
+type stateFn func(*lexer) stateFn
+
+// slashState is the first state before a segment
+func slashState(l *lexer) stateFn {
+	if !l.ignoreIfMatches(slash) {
+		return l.unexpectedRuneError(slash, l.peek())
+	}
+	return segmentState
+}
+
+func segmentState(l *lexer) stateFn {
+	switch l.peek() {
+	case eof, slash, verbSep:
+		return l.errorf("expected a segment, found %q", l.peek())
+	case varLeft:
+		return varState
+	}
+	return literalState
+}
+
+const validFirstIdentifierRunes = "ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
+	"abcdefghiljkmnopqrstuvwxyz"
+
+const validIdentifierRunes = validFirstIdentifierRunes +
+	"0123456789" +
+	"_"
+const validVariableValueRunes = validIdentifierRunes + "/*"
```

Comment:
I do not think `/` or `*` are ever valid variable runes.  I reckon identifiers are supposed to be protobuf field (and subfield) identifiers, which do not include these characters.

Maybe this helps:

https://github.com/googleapis/google-cloud-cpp/blob/4174d656136f4b849c8a3d327237f3a96be3e003/generator/internal/http_annotation_parser.h#L60-L74

Reply:
That helps a lot! Thank you! And [this interpretation](https://github.com/googleapis/google-cloud-cpp/blob/4174d656136f4b849c8a3d327237f3a96be3e003/generator/internal/http_annotation_parser.h#L49-L58) of the spec also answers a few questions I had.

Part of my confusion here is that the interface I am working with to express the parsed result doesn't seem to accommodate the recursive elements in the spec (it's just an array of strings that are classified as one of literal|variable|verb). I see the C++ parser has [a much richer data structure](https://github.com/googleapis/google-cloud-cpp/blob/4174d656136f4b849c8a3d327237f3a96be3e003/generator/internal/http_annotation_parser.h#L80-L95). I will model the parser after that and see how far I go.

Reply:
> 
> Part of my confusion here is that the interface I am working with to express the parsed result doesn't seem to accommodate the recursive elements in the spec (it's just an array of strings that are classified as one of literal|variable|verb).

Ah, yes.  Before changing [`PathInfo`](https://github.com/googleapis/google-cloud-rust/blob/9da4eceaec8777fab859c3d14f49c909f17090ac/generator/internal/api/model.go#L130), consider building a more complex data structure and try to flatten that into a `PathInfo`.

If the flattening fails we can exit the generator with an error (`slog.Fatal()` or `os.Exit(1)` or `panic()` whatever is idiomatic in golang).  On that day we can figure out what is the right way to handle the more complex case.

At the very least that gives you a way to keep the scope of these changes small.  We can replace `PathInfo` in a second PR if you become interested in that too.